### ISSUE TITLE
Multiple instances of node in ur program not possible 

### ISSUE THREAD

**Issue by ipa-lth** [UniversalRobots/Universal_Robots_ExternalControl_URCap#5]
With the current implementation it is not possible to have this urcap node more than once in the ur program.

The error message says:
Contextual error: The function set_servo_setpoint is double defined.

This function might be defined in the respective ur-script on the driver side.
However, I imaging that the function itself should be defined in the "installation" part instead of the "node" part.

**Comment by fmauch** [comment]
Related: #2

**Comment by fmauch** [comment]
The main problem is basically, that we would like to request the program from the ROS machine during runtime. However, if we do obtain it inside the installation part, then retrieving the program would be run every time once the URCap is installed, also when not being part of the program tree.

Probably, one possible solution I had in mind earlier was to define everything inside a function (Is it possible to define functions inside URScript functions?) and name that function with a unique identifier coming from the program tree. However, I never found the time to investigate this further.

**Comment by ipa-lth** [comment]
Hmmm, I see.
But spamming the program code with differently named functions is also not ideal.

Would it be possible to establish two communication channels in the urcap: First during "Installation" contribution and add some functions, than second in the "node" contribution to configure the function calls?

**Comment by fmauch** [comment]
I fully agree. Which is why I have been postponing this so far :-)

Ideally we could check, whether the corresponding ProgramNode is part of the program tree and then receive the code during the installation part. Unfortunately, to my knowledge, this doesn't exist in the API.

**Comment by ipa-lth** [comment]
I actually think that it would not be that hard:
[Here](https://github.com/UniversalRobots/Universal_Robots_ExternalControl_URCap/blob/master/src/main/java/com/fzi/externalcontrol/impl/ExternalControlInstallationNodeContribution.java#L69) the ur program is requested from the server and stored in the data-model. If we could seperate the "installation" part of the response, we could run the according ur-script already here.
In [the node part](https://github.com/UniversalRobots/Universal_Robots_ExternalControl_URCap/blob/master/src/main/java/com/fzi/externalcontrol/impl/ExternalControlProgramNodeContribution.java#L83) the response is then added to the urscript. Here only the node pieces would be added.

My biggest issue in not just post a PR is, that it is so tightly coupled with waht the ur_driver sends. And I haven't looked into the text (urscript) generation in the ur_driver,yet.

**Comment by ipa-lth** [comment]
I have a prototype on how to enable multiple nodes: #6 
I added some lines in the urcap to "uniquify" the received code.

Unfortunately, constant definitions in the received code are not resolved.
So, to make it work with the ur_driver every CONSTANT must be replace before sending.
I did this lokally with my texteditor, but haven't pushed to ur_driver, yet. (And probably won't, if not requested)

**Comment by gavanderhoorn** [comment]
> So, to make it work with the ur_driver

Nit: this should probably be `ur_robot_driver`.

`ur_driver` is the deprecated (and soon to be removed) Python-based initial driver which was part of `ros-industrial/universal_robot`.

For future readers, referring to `ur_driver` here will be very confusing.

**Comment by ipa-lth** [comment]
> 
> 
> > So, to make it work with the ur_driver
> 
> Nit: this should probably be `ur_robot_driver`.
> 
> `ur_driver` is the deprecated (and soon to be removed) Python-based initial driver which was part of `ros-industrial/universal_robot`.
> 
> For future readers, referring to `ur_driver` here will be very confusing.

Fixed

**Comment by fmauch** [comment]
> Unfortunately, constant definitions in the received code are not resolved.
> So, to make it work with the ur_robot_driver every CONSTANT must be replace before sending.
> I did this lokally with my texteditor, but haven't pushed to ur_robot_driver, yet. (And probably won't, if not requested)

I don't fully understand that comment. Which CONSTANTs are you referring to?

**Comment by ipa-lth** [comment]
> 
> 
> > Unfortunately, constant definitions in the received code are not resolved.
> > So, to make it work with the ur_robot_driver every CONSTANT must be replace before sending.
> > I did this lokally with my texteditor, but haven't pushed to ur_robot_driver, yet. (And probably won't, if not requested)
> 
> I don't fully understand that comment. Which CONSTANTs are you referring to?

E.g. this https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/resources/ros_control.urscript#L10

And also this: https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/resources/ros_control.urscript#L3

**Comment by fmauch** [comment]
And those can't be defined inside the functions? I would assume, that you can simply define them inside the function and they live inside the function's scope.

I would have been more concerned about the global variables as in https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/resources/ros_control.urscript#L20

**Comment by ipa-lth** [comment]
Global variable seem to work, but I haven't tested it some much yet.
I will report as soon as I know more

**Comment by ipa-lth** [comment]
> 
> 
> And those can't be defined inside the functions? I would assume, that you can simply define them inside the function and they live inside the function's scope.

They are available in the function scope but not in the scope of the function within the function... e.g. [here](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/resources/ros_control.urscript#L30)

**Comment by ipa-lth** [comment]
@fmauch @gavanderhoorn @BomMadsen 
I want to discuss with you how to proceed. My idea would be the following:

In ur_robot_driver: 

- Seperate the "PROGRAM_REQUEST" function into two calls, one for "Installation contribution" and one for "node contribution".
- Seperate the "ros_control.urscript" into two sections respectively

In the urcap:

- Call both requests during "installation-generation time" and apply the "installation" part. Store the other for later
- During the "node-generation time", each instance (of node contribution) is wrapped into a unique function name




+:

- the function of the urcap is more generalized for other approaches to use
- the new urCap would probably work with an old (current) version of the ur_robot_driver (and only state a warning of an unexpected request)


-: 

- more complexity must be introduced into the ur_robot_driver


Do you support such an approach? If I PR, would you merge it? Is there someone also working on this?

**Comment by gavanderhoorn** [comment]
Seeing as he's the author, I'll defer to @fmauch here and to @BomMadsen as the URCap expert.

**Comment by fmauch** [comment]
Let me come back on that tomorrow.

**Comment by fmauch** [comment]
Thanks for summing things up @ipa-lth 

While such an approach definitely would be more powerful, I don't see a direct advantage, especially regarding the original issue.

Having program requests inside the Installation Node Contribution would bring us back to #2 where you proposed two different workarounds for.

I'm just asking myself, why we then should pull something else inside the Program Node Contribution. This would only make sense if we pulled different code in each node contribution, right? That means we also would have to provide a selection of different code contributions back in the driver.
This would add the additional requirement to synchronize available scripts between the TP and the driver, while currently, there  is only one script that is passed to the driver on ROS side.

With leaving everything inside the installation part, we could run the current structure totally fine with simply calling the function in the Program Node contribution.

**Comment by ipa-lth** [comment]
I cannot follow completely.
Right now, the program is requested during Installation Node contribution and (fully) applied (adding it to the ur-program) at Program Node Contribution.

I propose to request two seperate code contributions. Both at Installation Node contribution.
One of these contributions would be applied directly (during installation node contribution). This part would exist only once in the resulting ur-program and should contain definitions of functions, threads etc, and possibly some initialization code.

Every time the program node contribution is entered in the following, an "activation" code snippet is added to the resulting ur-program. When designed well, this code is essentially a call to a previously defined function. However, to relieve the user from the burden to care that the node contribution (variable names etc) is unique, for allowing multiple instances in one ur-program, we additional wrap the received program node contribution snipped in a unique function from within the urcap.

I have the feeling, that it is a cleaner approach to seperated the two "program requests" to get the snippets. But both request will be called direclty after each other during installation node contribution.
Another possibility could be to somehow annotate the two distinct code snippets, send it as one string and then disect when received in the urcap.


> 
> Having program requests inside the Installation Node Contribution would bring us back to #2 where you proposed two different workarounds for.
> 
> ...
> 
> With leaving everything inside the installation part, we could run the current structure totally fine with simply calling the function in the Program Node contribution.

Unfortunetly, it does not work fine with the current ur_robot_driver. The [constant definitions](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/resources/ros_control.urscript#L9) must then either be global or resolved otherwise. Forcing "global" gives me some goose bumps, and I haven't tried it yet. But it could be a simple fix.

What do you think? I really want to resolve this issue, preferably with an approach you want to merge.

**Comment by ipa-lth** [comment]
Quoting form the scriptmanual_en.pdf, section 1.7 (p. 6):
>Every variable declared inside a program exits at a global scope, except when they
are declared inside a function. In that case the variable are local to that function. Two
qualifiers are available to modify this behaviour. The local qualifier tells the runtime to
treat a variable inside a function, as being truly local, even if a global variable with the
same name exists. The global qualifier forces a variable declared inside a function,
to be globally accessible.

In fact, adding a "global" statement just explicitly states the same functionality. There seems to be a slight difference though, since "globals" are visible in the "Variables" tab of the urcap.

**Comment by fmauch** [comment]
I think, that gave me a clearer understanding of your concerns. I agree that this might be a good approach, but we should take care to keep things as simple as possible.

> Unfortunetly, it does not work fine with the current ur_robot_driver. The [constant definitions](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/resources/ros_control.urscript#L9) must then either be global or resolved otherwise. Forcing "global" gives me some goose bumps, and I haven't tried it yet. But it could be a simple fix.

I'm afraid, as they will be used inside different functions, we probably will have to declare them globally. Note, that defining a variable `global` is something different than defining them outside of any function, as `global` variables will show up on the TP under the "Variables tab".

Basically, everything above https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/resources/ros_control.urscript#L96 will have to be written from the Installation Node Contribution, while everything below should come from the Program Node Contribution. It might also make sense to wrap all that code into another function to the preamble and then only call that function from the Program Node Contribution.

**Comment by fmauch** [comment]
Note: This comment is spanning a bit over this issue, #7 and https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/207

I've been thinking about this a bit and I do see benefits of writing everything into a separate function (as in #7) rather than populating the preamble with this:
 - When using multiple instances, the global variables are set inside the function to the value they are needed at inside that function. (Though still being global variables as in https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/207).
 - We don't have to split the code into a preamble part and a program node part
 - With that approach we could easily extend the URCap to support multiple ROS remotes in one TP program (There have been requests of doing this in the past)
 - We could move the code retrieval into the Program Node completely. (I think). This would make the workarounds mentioned in https://github.com/UniversalRobots/Universal_Robots_ExternalControl_URCap/issues/5#issuecomment-647451184 unnecesary.

The downside I see is that we populate the variable view with many new global variables as https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/207 introduces.

I'll check your implementation from #7 later today and then give feedback on that again.

**Comment by ipa-lth** [comment]
> The downside I see is that we populate the variable view with many new global variables as [UniversalRobots/Universal_Robots_ROS_Driver#207](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/207) introduces.

You are absolutely right. But to be honest, those have been global variables all along...

**Comment by fmauch** [comment]
> > The downside I see is that we populate the variable view with many new global variables as [UniversalRobots/Universal_Robots_ROS_Driver#207](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/207) introduces.
> 
> You are absolutely right. But to be honest, those have been global variables all along...

Yes, I am aware of that (and always was). My comment focuses on the variable view. Currently, we use the `global` keyword explicitly on the variables that we want to be shown in the variable view.

**Comment by ipa-lth** [comment]
> 
> 
> > > The downside I see is that we populate the variable view with many new global variables as [UniversalRobots/Universal_Robots_ROS_Driver#207](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/207) introduces.
> > 
> > 
> > You are absolutely right. But to be honest, those have been global variables all along...
> 
> Yes, I am aware of that (and always was). My comment focuses on the variable view. Currently, we use the `global` keyword explicitly on the variables that we want to be shown in the variable view.

Yes, that cluttering of the view is unfortunate.

**Comment by urrsk** [comment]
Thanks for the issue addressing this challenge.
#14 is able to handle multiple nodes. Thus to make a more robust solution, we need to split the URScript code in a more explicit way.

**Comment by fmauch** [comment]
I'm closing this, since #14 is merged since a while.
